File: internal/server/auth/password_strength.go
package auth

// password_strength.go — Phase 6.8 strength policy for new passwords.
//
// Applied to signup and password-reset-complete. Rejects passwords
// that are too short, too monotonous (one character class only), or
// that match a small blacklist of obvious entries (the username, the
// email local-part, "password", repeated digits, etc.).
//
// We deliberately stay closer to NIST SP 800-63B's "length over
// complexity" guidance than the bank-portal habit of "must contain
// exactly one of every category" — the policy gates the WORST cases
// (a 10-char string of digits, "passwordpassword") without
// micromanaging good entries.

import (
	"errors"
	"strings"
	"unicode"
)

const (
	// MinPasswordLen is the floor — 10 chars. Below this we reject
	// outright before checking anything else.
	MinPasswordLen = 10

	// MinPasswordCharClasses is how many of the four character
	// classes (lower, upper, digit, symbol) a password must mix.
	// Two of four (e.g. lowercase + digit) is enough to keep the
	// keyspace honest without locking out users who can't type a
	// symbol on every keyboard.
	MinPasswordCharClasses = 2

	// MaxConsecutiveRepeat is how many times the same character can
	// repeat in a row before we reject. Catches "aaaaaaaaaa" /
	// "1111111111" without flagging legitimate doubles like "letter".
	MaxConsecutiveRepeat = 4
)

// ErrWeakPassword is the single sentinel surfaced to callers. The
// detailed reason is logged but not returned to the client — telling
// an attacker exactly which rule their guess violated narrows their
// search.
var ErrWeakPassword = errors.New("weak password")

// commonWeakPasswords is a tiny blacklist of patterns that show up
// over and over in credential dumps. Lowercased for comparison.
// Intentionally short: a real defense uses HIBP / a breach API; this
// is just enough to refuse the most embarrassing entries before
// argon2id even runs.
var commonWeakPasswords = map[string]struct{}{
	"password":      {},
	"password1":     {},
	"password12":    {},
	"qwerty":        {},
	"qwerty123":     {},
	"qwertyuiop":    {},
	"1234567890":    {},
	"abcdefghij":    {},
	"letmein":       {},
	"welcome":       {},
	"admin":         {},
	"administrator": {},
	"changeme":      {},
	"iloveyou":      {},
	"floffi":        {},
	"admin12345":    {},
}

// CheckPasswordStrength validates a candidate password against the
// project policy. usernameAndEmail lets the caller pass identifiers
// that must NOT appear inside the password (so "alice123" can't
// sign up with password "alice123456"). Pass empty strings if no
// identifiers are known yet.
//
// Returns nil on accept, ErrWeakPassword otherwise.
func CheckPasswordStrength(password, username, email string) error {
	if len(password) < MinPasswordLen {
		return ErrWeakPassword
	}

	// Character-class count.
	var hasLower, hasUpper, hasDigit, hasSymbol bool
	var lastRune rune
	var repeatCount int
	for i, r := range password {
		switch {
		case unicode.IsLower(r):
			hasLower = true
		case unicode.IsUpper(r):
			hasUpper = true
		case unicode.IsDigit(r):
			hasDigit = true
		case unicode.IsSpace(r):
			// Allow internal whitespace (passphrases) but it doesn't
			// count toward any class.
		default:
			// Anything else (punctuation, symbols, unicode marks).
			hasSymbol = true
		}

		// Track consecutive-same-rune runs. i==0 means lastRune is
		// the zero value; reset rather than count against it.
		if i > 0 && r == lastRune {
			repeatCount++
			if repeatCount >= MaxConsecutiveRepeat {
				return ErrWeakPassword
			}
		} else {
			repeatCount = 1
		}
		lastRune = r
	}

	classes := 0
	for _, b := range []bool{hasLower, hasUpper, hasDigit, hasSymbol} {
		if b {
			classes++
		}
	}
	if classes < MinPasswordCharClasses {
		return ErrWeakPassword
	}

	// Common-password blacklist (case-insensitive).
	lc := strings.ToLower(password)
	if _, bad := commonWeakPasswords[lc]; bad {
		return ErrWeakPassword
	}

	// Username / email local-part containment. A password that's a
	// trivial expansion of the user's handle ("alice2024") gives up
	// most of its entropy to anyone who knows the account.
	//
	// Minimum identifier length is 5 — anything shorter (rec, ab,
	// dev, etc.) collides with normal English fragments inside
	// otherwise-strong passwords ("correct-horse" trips on a 3-char
	// "rec" from the email). Five chars is roughly where an
	// identifier stops looking like a generic word fragment.
	const minIdentifierLen = 5
	if len(username) >= minIdentifierLen {
		if strings.Contains(lc, strings.ToLower(username)) {
			return ErrWeakPassword
		}
	}
	if email != "" {
		if at := strings.IndexByte(email, '@'); at > 0 {
			local := strings.ToLower(email[:at])
			if len(local) >= minIdentifierLen && strings.Contains(lc, local) {
				return ErrWeakPassword
			}
		}
	}

	return nil
}

File: frontend/src/lib/passwordStrength.ts
// passwordStrength — client-side mirror of the server policy in
// internal/server/auth/password_strength.go. The server is the
// source of truth (a stale client can't bypass it), but mirroring
// the rules here lets us show the user a precise reason before
// any round-trip.

export const MIN_PASSWORD_LEN = 10;
export const MIN_CHAR_CLASSES = 2;
export const MAX_CONSECUTIVE_REPEAT = 4;
export const MIN_IDENTIFIER_LEN = 5;

const COMMON_WEAK = new Set([
  "password", "password1", "password12",
  "qwerty", "qwerty123", "qwertyuiop",
  "1234567890", "abcdefghij",
  "letmein", "welcome",
  "admin", "administrator",
  "changeme", "iloveyou",
  "floffi", "admin12345",
]);

export type PasswordIssue =
  | "too_short"
  | "needs_more_classes"
  | "too_repetitive"
  | "common_password"
  | "contains_identifier";

export function checkPasswordStrength(
  password: string,
  username = "",
  email = "",
): PasswordIssue | null {
  if (password.length < MIN_PASSWORD_LEN) return "too_short";

  // Character class detection mirrors Go's unicode.IsLower/Upper/Digit.
  // "Symbol" is the catch-all for anything that isn't a letter or digit
  // (whitespace doesn't count toward any class).
  let lower = false, upper = false, digit = false, symbol = false;
  let lastChar = "";
  let runLen = 0;
  for (let i = 0; i < password.length; i++) {
    const ch = password[i];
    if (/[a-zß-öø-ÿ]/.test(ch) || (ch >= "a" && ch <= "z")) lower = true;
    else if (/[A-ZÀ-ÖØ-Þ]/.test(ch) || (ch >= "A" && ch <= "Z")) upper = true;
    else if (ch >= "0" && ch <= "9") digit = true;
    else if (ch !== " " && ch !== "\t") symbol = true;
    // Non-Latin letters (e.g. Korean) also land in "symbol" via this
    // branch, which matches the Go side's default-case bucket — they
    // legitimately add entropy.

    if (i > 0 && ch === lastChar) {
      runLen++;
      if (runLen >= MAX_CONSECUTIVE_REPEAT) return "too_repetitive";
    } else {
      runLen = 1;
    }
    lastChar = ch;
  }

  const classes = (lower ? 1 : 0) + (upper ? 1 : 0) + (digit ? 1 : 0) + (symbol ? 1 : 0);
  if (classes < MIN_CHAR_CLASSES) return "needs_more_classes";

  const lc = password.toLowerCase();
  if (COMMON_WEAK.has(lc)) return "common_password";

  if (username.length >= MIN_IDENTIFIER_LEN && lc.includes(username.toLowerCase())) {
    return "contains_identifier";
  }
  if (email) {
    const at = email.indexOf("@");
    const local = (at > 0 ? email.slice(0, at) : email).toLowerCase();
    if (local.length >= MIN_IDENTIFIER_LEN && lc.includes(local)) {
      return "contains_identifier";
    }
  }
  return null;
}

// gradePasswordStrength — coarse UX score on top of checkPasswordStrength.
// The policy gate already decides pass/fail; this just maps the user's
// input to a 4-level meter so they get a "stronger" signal as they keep
// adding entropy past the minimum bar.
//
// Levels:
//   0 약함        — doesn't satisfy the policy yet (has a PasswordIssue)
//   1 보통        — passes the policy but only barely (short + 2 classes)
//   2 강함        — passes + 12자 이상 with 3 classes, or 14자+ with 2
//   3 매우 강함   — passes + 16자 이상 with 3+ classes, or 12자+ with all 4
//
// Returns null for empty input so the caller can hide the meter entirely.
export type PasswordGrade = {
  level: 0 | 1 | 2 | 3;
  label: string;
};

export function gradePasswordStrength(
  password: string,
  username = "",
  email = "",
): PasswordGrade | null {
  if (password.length === 0) return null;

  if (checkPasswordStrength(password, username, email) !== null) {
    return { level: 0, label: "약함" };
  }

  // Count character classes the same way the gate does.
  let lower = false, upper = false, digit = false, symbol = false;
  for (let i = 0; i < password.length; i++) {
    const ch = password[i];
    if (/[a-zß-öø-ÿ]/.test(ch) || (ch >= "a" && ch <= "z")) lower = true;
    else if (/[A-ZÀ-ÖØ-Þ]/.test(ch) || (ch >= "A" && ch <= "Z")) upper = true;
    else if (ch >= "0" && ch <= "9") digit = true;
    else if (ch !== " " && ch !== "\t") symbol = true;
  }
  const classes = (lower ? 1 : 0) + (upper ? 1 : 0) + (digit ? 1 : 0) + (symbol ? 1 : 0);
  const len = password.length;

  if ((len >= 16 && classes >= 3) || (len >= 12 && classes >= 4)) {
    return { level: 3, label: "매우 강함" };
  }
  if ((len >= 12 && classes >= 3) || (len >= 14 && classes >= 2)) {
    return { level: 2, label: "강함" };
  }
  return { level: 1, label: "보통" };
}

// Human-readable Korean message for each issue. Used by SignupView /
// ResetPasswordView so the error banner says what's wrong, not just
// "weak password."
export function describePasswordIssue(issue: PasswordIssue): string {
  switch (issue) {
    case "too_short":
      return `비밀번호는 최소 ${MIN_PASSWORD_LEN}자 이상이어야 해요.`;
    case "needs_more_classes":
      return "영문 대/소문자, 숫자, 기호 중 두 가지 이상을 섞어주세요.";
    case "too_repetitive":
      return `같은 문자가 ${MAX_CONSECUTIVE_REPEAT}번 이상 연속될 수 없어요.`;
    case "common_password":
      return "너무 흔한 비밀번호예요. 다른 조합을 사용해주세요.";
    case "contains_identifier":
      return "비밀번호에 아이디나 이메일 일부를 포함할 수 없어요.";
  }
}
